dftimewolf/lib/containers/manager.py

```diff
@@ -0,0 +1,137 @@
+"""A ContainerManager class."""
+
+
+import dataclasses
+import threading
+import typing
+from typing import cast, Sequence, Type
+
+from dftimewolf.lib.containers import interface
+
+
+@dataclasses.dataclass
+class _MODULE():
+  name: str
+  dependencies: list[str] = dataclasses.field(default_factory=list)
+  storage: list[interface.AttributeContainer] = dataclasses.field(
+      default_factory=list)
+  completed: bool = False
+
+
+class ContainerManager():
+  """A ContainerManager.
+  
+  This ContainerManager handles container storage and delivery to modules.
+  Modules can only receive containers from other modules that they directly
+  depend on, according to the recipe (or themselves.) In this way, it implements
+  a directional graph for container delivery.
+
+  Attributes:
+    _mutex: Practice safe container access.
+    _modules: Container storage and dependency information.
+  """
+
+  def __init__(self) -> None:
+    """Initialise a ContainerManager."""
+    self._mutex = threading.Lock()
+    self._modules: dict[str, _MODULE] = {}
+
+  def ParseRecipe(self, recipe: dict[str, typing.Any]) -> None:
+    """Parses a recipe to build the dependency graph."""
+    with self._mutex:
+      self._modules = {}
+
+      for module in recipe.get('preflights', []) + recipe.get('modules', []):
+        name = module.get('runtime_name', module.get('name', None))
+        if not name:
+          raise RuntimeError("Name not set for module in recipe")
+
+        self._modules[name] = _MODULE(
+            name=name, dependencies=module.get('wants', []) + [name])
+
+  def StoreContainer(self,
+                     source_module: str,
+                     container: interface.AttributeContainer) -> None:
+    """Adds a container to storage for later retrieval.
+
+    Args:
+      source_module: The module that generated the container.
+      container: The container to store.
+    """
+    if len(self._modules) == 0:
```

Comment:
you could probably just use the fact an empty sequence evaluates as False?

Reply:
Done.